### PR TITLE
Honor component-level attributes in dynamic version selection

### PR DESCRIPTION
**Pull request by melix**
### Context

This pull request implements the missing bit in attribute matching aware selection: honoring component-level attributes during dynamic version selection.

Before this commit, component-level attributes were ignored whenever a dynamic version selector was used (eg: `1.+`, `[1.0, )`, `latest.release`, ...). This resulted in an inconsistent behavior, where a version may be selected by the dynamic version resolver (`latest.release` resolves to 1.3) but then evicted during _variant_ selection, when an earlier release might have passed.

The side effect of this PR is that _all_ dynamic selectors will trigger the resolution (downloading) of component metadata for a pre-selected module as soon as it is pre-selected. This is because component level attributes are found on the component metadata, which in turns requires downloading metadata. Before, only `latest.release` selector required this, because the "status" was found in metadata.

In short, this may have an impact on performance, in addition to the extra matching.

### CODE REVIEW

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ComponentAttributesDynamicVersionIntegrationTest.groovy [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+            }
+            'org.test:module:1.0' {
+                attribute('quality', 'beta')
+            }
+        }
+        buildFile << """
+            def quality = Attribute.of("quality", String)
+
+            configurations {
+                // This test also makes sure that configuration-level attributes are overwritten
```
Comment (ljacomet) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ComponentAttributesDynamicVersionIntegrationTest.groovy]: Looking at the comment, I tracked that override logic - unless mistaken - into `DefaultImmutableAttributesFactory.concat` but did not see a unit test explicitly showing that behavior.
And I also noticed that it is missing tests for the recently added `safeConcat`.
These two things might be worth covering in unit tests.

Do not think it should block merging this though, but a follow up is needed.

**Comment by oehme** [comment]
Just commenting on the potential performance impact: Is the metadata always necessary? Or only when dependency-level attributes are used?

**Comment by melix** [comment]
Metadata is always required, because you have _configuration_ attributes too. So far they weren't used during component resolution, so we could select a version that doesn't match the attributes in the range. But, as I said, this is only for _dynamic selectors_. For fixed versions, there's no impact.

**Comment by oehme** [comment]
Yes, though dynamic versions are (for better or worse) widely used. I guess you can mitigate the impact by using locking? Does that reduce the search space and downloads?
